File: WnW_Base/src/main/java/de/jjl/wnw/base/util/path/WnWPoint.java
package de.jjl.wnw.base.util.path;


public class WnWPoint
{
	private int x;
	
	private int y;
	
	public static final WnWPoint ZERO = new WnWPoint(0, 0);

	public WnWPoint(int x, int y)
	{
		this.x = x;
		this.y = y;
	}
	
	@Override
	public boolean equals(Object obj)
	{
		if(obj == this)
		{
			return true;
		}
		
		if(!(obj instanceof WnWPoint))
		{
			return false;
		}
		
		return x == ((WnWPoint) obj).x && y == ((WnWPoint) obj).y;
	}
	
	@Override
	public String toString()
	{
		return "WnWPoint [" + x + "," + y + "]";
	}
	
	public WnWPoint negatePoint()
	{
		return new WnWPoint(-x, -y);
	}
	
	public WnWPoint plus(WnWPoint b)
	{
		return new WnWPoint(x + b.x, y + b.y);
	}
	
	public WnWPoint minus(WnWPoint b)
	{
		return new WnWPoint(x - b.x, y - b.y);
	}
	
//	public WnWPoint div(WnWPoint b)
//	{
//		return new WnWPoint(x / b.x, y / b.y);
//	}
	
	public WnWPoint div(int factor)
	{
		return new WnWPoint(x / factor, y / factor);
	}
	
	public WnWPoint rem(int factor)
	{
		return new WnWPoint(x % factor, y % factor);
	}

	public int getY()
	{
		return y;
	}
	
	public int getX()
	{
		return x;
	}

	public WnWPoint times(int factor)
	{
		return new WnWPoint(x * factor, y * factor);
	}

	public int length()
	{
		return this.x + this.y;
	}

	public WnWPoint div(WnWPoint b)
	{
		return new WnWPoint(x / b.x, y / b.y);
	}

	public WnWPoint times(WnWPoint b)
	{
		return new WnWPoint(x * b.x, y * b.y);
	}
}

File: WnW_Base/src/main/kotlin/de/jjl/wnw/base/util/path/WnWPoint.kt
package de.jjl.wnw.base.util.path

data class WnWPoint(val x: Int, val y: Int)
{
	companion object
	{
		@JvmStatic val ZERO = WnWPoint(0, 0)
	}
}

operator fun WnWPoint.unaryPlus() = this

operator fun WnWPoint.unaryMinus() = WnWPoint(-this.x, -this.y)

operator fun WnWPoint.plus(b: WnWPoint) = WnWPoint(x + b.x, y + b.y)

operator fun WnWPoint.minus(b: WnWPoint) = WnWPoint(x - b.x, y - b.y)

operator fun WnWPoint.times(factor: Int) = WnWPoint(x * factor, y * factor)

operator fun WnWPoint.div(factor: Int) = WnWPoint(x / factor, y / factor)

operator fun WnWPoint.rem(factor: Int) = WnWPoint(x % factor, y % factor)
